drivers/counter/counter_nrfx_rtc.c

```diff
@@ -155,12 +154,14 @@ static int counter_nrfx_set_top_value(struct device *dev, u32_t ticks,
 	}
 
 	nrfx_rtc_cc_disable(rtc, TOP_CH);
-	nrfx_rtc_counter_clear(rtc);
+	if (cfg->flags & COUNTER_TOP_FLAG_RESET) {
+		nrfx_rtc_counter_clear(rtc);
+	}
 
-	dev_data->top_cb = callback;
-	dev_data->top_user_data = user_data;
-	dev_data->top = ticks;
-	nrfx_rtc_cc_set(rtc, TOP_CH, ticks, callback ? true : false);
+	dev_data->top_cb = cfg->callback;
+	dev_data->top_user_data = cfg->user_data;
+	dev_data->top = cfg->ticks;
+	nrfx_rtc_cc_set(rtc, TOP_CH, cfg->ticks, cfg->callback ? true : false);
```

Comment:
This part is a bit tricky. User of the API may want to change the top value up and down by just a few counts. I.e. a PI controller would do exactly such a thing. `nrfx_rtc_cc_disable` disables channel events and channel interrupts. If in between the call to `nrfx_rtc_cc_disable` and `nrfx_rtc_cc_set` the count value would pass the new 'top value' we would fail to trigger the callback and have 'runaway' counter.

An example: 
- user calls `counter_set_top_value` when the count value is 900
- when `nrfx_rtc_cc_disable` is finally executed the counter has value 999
- when `nrfx_rtc_cc_set` sets the 'top value' to 1000 the count value is 1001

This is similar, though not identical to my last comment. From the application point of view neither decreasing nor increasing the top value is safe here. I believe we cannot claim it's an application responsibility to mitigate the issue. Application, working in a preemptive RTOS, has limited control over timing. For the user of the API it's going to be impossible to estimate a safe margin (unless it's a very rough estimation) when the `counter_set_top_value` can be called in relation to the current counter value.

One safe way a counter driver can implement the feature is to
- block all the interrupts
- read the current counter value
- knowing the frequency of the counter verify that we have enough time to set the new top value before counter reloads
1. if there is not enough time to set the new top value: enable interrupts, wait until the counter reloads itself, set the new top value
2. if there is enough time: set the new top value, enable interrupts

Not calling the  `nrfx_rtc_cc_disable` would be another option but it doesn't solve the problem when user wants to decrease the top value by just a few counts.

The `COUNTER_TOP_FLAG_RESET` may have been added to support kernel usage scenarios described in  #12068 where the above issue may not be relevant at all. This is part of the public API however. Other users may use the API differently.

Reply:
@mnkp user can do something like:

```
cfg.flags = 0; /* no reset */
counter_set_top_value(dev, &cfg);
if (counter_read(dev) > counter_get_top_value(dev)) {
    cfg.flags = COUNTER_TOP_FLAG_RESET;
    counter_set_top_value(dev, &cfg);
}
```
Of course, such counter will lose some ticks but that is inevitable if we missed setting it.

Alternatively, it can be inside `counter_set_top_value` and return error if counter exceeded top value.

Reply:
@mnkp any comments? do  you agree that it can stay as the user responsibility?

Reply:
This touches on areas of concern I have in #16252, i.e. no [clear model of what a counter is](https://github.com/zephyrproject-rtos/zephyr/pull/16252#issuecomment-494174959), but looking at this I can't imagine why we would expect an application to have to deal with a case where `counter_read()`  produced a value greater than `counter_get_top_value()`, no matter what was going on.  That's just not consistent with how a counter should behave, unless "top value" doesn't mean top value.

Reply:
So I have following proposal which would be similar to #16252: there is an additional flag which instructs driver to reset counter if top value if exceeded or return error. For certain scenarios lost ticks will be accepted and then user will set flag to reset and for other it may indicate that application has performance issues and error should be handled.

Reply:
@mnkp @pabigot what do you think?